crypto/signatures.go

```diff
@@ -69,8 +71,35 @@ func VerifyHash(data Hash, pk PublicKey, sig Signature) error {
 	return nil
 }
 
+// WriteSignedData writes a length-prefixed object followed by its signature.
+func WriteSignedData(w io.Writer, obj interface{}, sk SecretKey) error {
+	b := encoding.Marshal(obj)
+	sig, _ := SignHash(HashBytes(b), sk) // no error possible
+	return encoding.NewEncoder(w).EncodeAll(b, sig)
+}
```

Comment:
It doesn't stream things. It takes all of the objects into memory, encodes them in memory, hashes them in memory, and then writes them.

I think this function would make more sense and have utility as a non-reader that just returns a []byte. Then the function could be used in non-reader scenarios like in transaction validation.